safehttp/plugins/collector/collector.go

```diff
@@ -0,0 +1,236 @@
+			return w.ClientError(safehttp.StatusBadRequest)
+		}
+
+		ct := r.Header.Get("Content-Type")
+		if ct == "application/csp-report" && hb.CSPHandler != nil {
+			return handleDeprecatedCSPReports(hb.CSPHandler, w, b)
+		} else if ct == "application/reports+json" && hb.Handler != nil {
+			return handleReport(hb.Handler, w, b)
+		}
+
+		return w.ClientError(safehttp.StatusUnsupportedMediaType)
+	})
+}
+
+func stringOrEmpty(x interface{}) string {
+	s, ok := x.(string)
+	if !ok {
+		return ""
+	}
+	return s
+}
+
+func uintOrZero(x interface{}) uint {
+	// x must be a float64 that will be converted to a uint since this is how json stores
+	// its numbers.
+	f, ok := x.(float64)
+	if !ok || f < 0 {
+		return 0
+	}
+	return uint(f)
+}
+
+func handleDeprecatedCSPReports(h func(CSPReport), w *safehttp.ResponseWriter, b []byte) safehttp.Result {
+	// In CSP2 it is clearly stated that a report has a single key 'csp-report'
+	// which holds the report object. Like this:
+	// {
+	//   "csp-report": {
+	//     // report goes here
+	//   }
+	// }
+	// Source: https://www.w3.org/TR/CSP2/#violation-reports
+	//
+	// But in the CSP3 spec this 'csp-report' key is never mentioned. So the report
+	// would look like this:
+	// {
+	//   // report goes here
+	// }
+	// Source: https://w3c.github.io/webappsec-csp/#deprecated-serialize-violation
+	//
+	// Because of this we have to support both. :/
+	m := make(map[string]interface{})
+	if err := json.Unmarshal(b, &m); err != nil {
+		return w.ClientError(safehttp.StatusBadRequest)
+	}
+
+	if report, exists := m["csp-report"]; exists {
+		var ok bool
+		m, ok = report.(map[string]interface{})
+		if !ok {
+			return w.ClientError(safehttp.StatusBadRequest)
+		}
+	}
+
+	ln := uintOrZero(m["lineno"])
+	if ln == 0 {
+		ln = uintOrZero(m["line-number"])
+	}
+	cn := uintOrZero(m["colno"])
+	if cn == 0 {
+		cn = uintOrZero(m["column-number"])
+	}
+
+	r := CSPReport{
+		BlockedURL:         stringOrEmpty(m["blocked-uri"]),
+		Disposition:        stringOrEmpty(m["disposition"]),
+		DocumentURL:        stringOrEmpty(m["document-uri"]),
+		EffectiveDirective: stringOrEmpty(m["effective-directive"]),
+		OriginalPolicy:     stringOrEmpty(m["original-policy"]),
+		Referrer:           stringOrEmpty(m["referrer"]),
+		Sample:             stringOrEmpty(m["script-sample"]),
+		StatusCode:         uintOrZero(m["status-code"]),
+		ViolatedDirective:  stringOrEmpty(m["violated-directive"]),
+		SourceFile:         stringOrEmpty(m["source-file"]),
+		LineNumber:         ln,
+		ColumnNumber:       cn,
+	}
+	h(r)
+
+	return w.NoContent()
+}
+
+func handleReport(h func(Report), w *safehttp.ResponseWriter, b []byte) safehttp.Result {
+	var rList []Report
+	if err := json.Unmarshal(b, &rList); err != nil {
+		return w.ClientError(safehttp.StatusBadRequest)
+	}
+
+	badRequest := false
+	for _, r := range rList {
+		m, ok := r.Body.(map[string]interface{})
+		if !ok {
+			badRequest = true
+			continue
+		}
+
+		if r.Type == "csp-violation" {
```

Comment:
I think it would be beneficial to have a `map[string]func(json.RawMessage)interface{}` and if the lookup fails we just use `json.Unmarshal` with `map[string]interface{}`.